DetectorDescription/DDCMS/plugins/dd4hep/DDDefinitions2Objects.cc

```diff
@@ -1853,208 +1853,214 @@ void Converter<print_xml_doc>::operator()(xml_h element) const {
+      }
+      for (xml::Document d : res.includes) {
+        print_doc((doc = d).root());
+        xml_coll_t(d.root(), DD_CMU(LogicalPartSection)).for_each(Converter<LogicalPartSection>(det, &context));
+      }
+      for (xml::Document d : res.includes) {
+        print_doc((doc = d).root());
+        xml_coll_t(d.root(), DD_CMU(Algorithm)).for_each(Converter<DDLAlgorithm>(det, &context));
+      }
+      for (xml::Document d : res.includes) {
+        print_doc((doc = d).root());
+        xml_coll_t(d.root(), DD_CMU(PosPartSection)).for_each(Converter<PosPartSection>(det, &context));
+      }
+      for (xml::Document d : res.includes) {
+        print_doc((doc = d).root());
+        xml_coll_t(d.root(), DD_CMU(SpecParSection)).for_each(Converter<SpecParSection>(det, &context));
+      }
 
-    /// Unload all XML files after processing
-    for (xml::Document d : res.includes)
-      Converter<include_unload>(det, &context, &res)(d.root());
-
-    print_doc((doc = dddef.document()).root());
-    // Now process the actual geometry items
-    xml_coll_t(dddef, DD_CMU(SolidSection)).for_each(Converter<SolidSection>(det, &context));
-    {
-      // Before we continue, we have to resolve all shapes NOW!
-      // Note: This only happens in a legacy DB payloads where
-      // boolean shapes can be defined before thier
-      // component shapes
-
-      while (!context.unresolvedShapes.empty()) {
-        for (auto it = context.unresolvedShapes.begin(); it != context.unresolvedShapes.end();) {
-          auto const& name = it->first;
-          auto const& aname = std::visit([](auto&& arg) -> std::string { return arg.firstSolidName; }, it->second);
-          auto const& bname = std::visit([](auto&& arg) -> std::string { return arg.secondSolidName; }, it->second);
-
-          auto const& ait = context.shapes.find(aname);
-          if (ait->second.isValid()) {
-            auto const& bit = context.shapes.find(bname);
-            if (bit->second.isValid()) {
-              dd4hep::Solid shape = std::visit(
-                  [&ait, &bit](auto&& arg) -> dd4hep::Solid { return arg.make(ait->second, bit->second); }, it->second);
-              context.shapes[name] = shape;
-              it = context.unresolvedShapes.erase(it);
+      /// Unload all XML files after processing
+      for (xml::Document d : res.includes)
+        Converter<include_unload>(det, &context, &res)(d.root());
+
+      print_doc((doc = dddef.document()).root());
+      // Now process the actual geometry items
+      xml_coll_t(dddef, DD_CMU(SolidSection)).for_each(Converter<SolidSection>(det, &context));
+      {
+        // Before we continue, we have to resolve all shapes NOW!
+        // Note: This only happens in a legacy DB payloads where
+        // boolean shapes can be defined before thier
+        // component shapes
+
+        while (!context.unresolvedShapes.empty()) {
```

Comment:
Isn't there a danger of an infinite loop here? If `ait` or `bit` are invalid, then `context.unresolvedShapes.erase(it)` is never called, and `unresolvedShapes` never becomes empty.

Reply:
yes, there is. It was intended as a workaround, not a solution. It's clear that there is no need for handling unresolved shapes since it is possible to define them in order. We need to make sure the definition order is enforced before parsing XML and throw an exception here.

Reply:
"throw an exception here." Can this problem be fixed now in this PR, or does an issue have to be written up about it?

Reply:
I'd think, this is definitely a separate issue. It requires an implementation of an XML validation tool (either a script or an analyzer). Then an exception can be thrown here.

Reply:
and... also applying the validation changes to current XML files

Reply:
In the short term, this infinite loop needs a quick fix. I ran into it myself, so it is not a hypothetical problem. If it can't be fixed in this PR, I'll submit a PR with a fix.